tofu init/get fails when fetching modules from private module registry

### OpenTofu Version

```shell
1.8.2
```


### OpenTofu Configuration Files

```hcl
module "my_module" {
  source = "private-registry.example.com/namespace/module_name/system"
  version = "some_version_that_exists_in_the_registry"
}
```


### Debug Output

```
Initializing modules...
2024-10-16T12:31:36.656-0400 [TRACE] ModuleInstaller: installing child modules for . into .terraform/modules
2024-10-16T12:31:36.656-0400 [DEBUG] Module installer: begin aws_default_tags
2024-10-16T12:31:36.656-0400 [TRACE] ModuleInstaller: aws_default_tags is not yet installed
2024-10-16T12:31:36.656-0400 [TRACE] ModuleInstaller: cleaning directory .terraform/modules/aws_default_tags prior to install of aws_default_tags
2024-10-16T12:31:36.656-0400 [TRACE] ModuleInstaller: aws_default_tags is a registry module at terraform.<company>.com/infrastructure/aws_default_tags/<namespace>
2024-10-16T12:31:36.656-0400 [DEBUG] aws_default_tags listing available versions of terraform.<company>.com/<namespace>/aws_default_tags/<system>at terraform.<company>.com
2024-10-16T12:31:36.656-0400 [DEBUG] Service discovery for terraform.<company>.com at https://terraform.<company>.com/.well-known/terraform.json
2024-10-16T12:31:36.656-0400 [TRACE] HTTP client GET request to https://terraform.<company>.com/.well-known/terraform.json
2024-10-16T12:31:36.794-0400 [DEBUG] Service discovery redirected to https://<APIGW_ID>.execute-api.us-east-1.amazonaws.com/live/.well-known/terraform.json
2024-10-16T12:31:36.794-0400 [TRACE] HTTP client GET request to https://<APIGW_ID>.execute-api.us-east-1.amazonaws.com/live/.well-known/terraform.json
2024-10-16T12:31:36.972-0400 [DEBUG] fetching module versions from "https://<APIGW_ID>.execute-api.us-east-1.amazonaws.com/live/v1/modules/<namespace>/aws_default_tags/<system>/versions"
2024-10-16T12:31:36.972-0400 [DEBUG] GET https://<APIGW_ID>.execute-api.us-east-1.amazonaws.com/live/v1/modules/<namespace>/aws_default_tags/<system>/versions
2024-10-16T12:31:36.972-0400 [TRACE] HTTP client GET request to https://<APIGW_ID>.execute-api.us-east-1.amazonaws.com/live/v1/modules/<namespace>/aws_default_tags/<system>/versions
2024-10-16T12:31:37.078-0400 [DEBUG] found available version "1.0.0" for
2024-10-16T12:31:37.078-0400 [DEBUG] found available version "1.0.1" for
2024-10-16T12:31:37.078-0400 [DEBUG] found available version "1.0.2" for
2024-10-16T12:31:37.078-0400 [DEBUG] found available version "1.0.3" for
2024-10-16T12:31:37.078-0400 [DEBUG] found available version "1.0.4" for
2024-10-16T12:31:37.078-0400 [DEBUG] found available version "1.1.0" for
2024-10-16T12:31:37.078-0400 [DEBUG] found available version "1.1.1" for
2024-10-16T12:31:37.078-0400 [DEBUG] found available version "1.1.2" for
2024-10-16T12:31:37.079-0400 [DEBUG] found available version "1.1.3" for
2024-10-16T12:31:37.079-0400 [DEBUG] found available version "1.1.4" for
Downloading terraform.<company>.com/infrastructure/aws_default_tags/<namespace>1.1.4 for aws_default_tags...
2024-10-16T12:31:37.079-0400 [DEBUG] looking up module location from "https://<APIGW_ID>.execute-api.us-east-1.amazonaws.com/live/v1/modules/<namespace>/aws_default_tags/<system>/1.1.4/download"
2024-10-16T12:31:37.079-0400 [DEBUG] GET https://<APIGW_ID>.execute-api.us-east-1.amazonaws.com/live/v1/modules/<namespace>/aws_default_tags/<system>/1.1.4/download
2024-10-16T12:31:37.079-0400 [TRACE] HTTP client GET request to https://<APIGW_ID>.execute-api.us-east-1.amazonaws.com/live/v1/modules/<namespace>/aws_default_tags/<system>/1.1.4/download
2024-10-16T12:31:37.141-0400 [ERROR] aws_default_tags from terraform.<company>.com/infrastructure/aws_default_tags/<namespace>1.1.4: failed to get download URL for "terraform.<company>.com/infrastructure/aws_default_tags/<system>": 200 OK resp:{"Item":{"id":{"S":"<namespace>/aws_default_tags/<system>"},"version":{"S":"1.1.4"},"source":{"S":"https://terraform-registry-<company>-<account_id>.s3.us-east-1.amazonaws.com/modules/<namespace>/aws_default_tags/<system>/1.1.4/aws_default_tags.zip"}}}
```

### Expected Behavior

init and get should "just work" and download the module artifacts from the S3 URL specified in the response from the registry API.

### Actual Behavior

```
│ Error: Error accessing remote module registry
│ 
│ Failed to retrieve a download URL for terraform.<company>.com/<namespace>/<module_name>/<system> <version> from terraform.<company>.com: failed to get download URL for
│ "terraform.<company>.com/<namespace>/<module_name>/<system>": 200 OK
│ resp:{"Item":{"id":{"S":"<namespace>/<module_name>/<system>"},"version":{"S":"<version>"},"source":{"S":"https://terraform-registry-<company>-<aws_account>.s3.us-east-1.amazonaws.com/modules/<namespace>/<module_name>/<system>/<version>/<module_name>.zip"}}}
```

### Steps to Reproduce

`rm -fr .terraform && tofu init` OR
`rm -fr .terraform && tofu get`

### Additional Context

This works when reverting to terraform 1.5.7 (1.6.3 also works, I assume all terraform versions will work).

`rm -fr .terraform && terraform init` will successfully discover and fetch the providers and modules from the private registry with no errors.

`rm -fr .terraform && tofu init` will result in the behavior described in this issue.

Since this fails at the module installation step I have not been able to verify if fetching providers will work. 

### References

_No response_

--- a/internal/registry/client.go
+++ b/internal/registry/client.go
@@ -22,6 +22,7 @@ import (
 	"github.com/hashicorp/go-retryablehttp"
 	svchost "github.com/hashicorp/terraform-svchost"
 	"github.com/hashicorp/terraform-svchost/disco"
+
 	"github.com/opentofu/opentofu/internal/httpclient"
 	"github.com/opentofu/opentofu/internal/logging"
 	"github.com/opentofu/opentofu/internal/registry/regsrc"
@@ -245,6 +246,11 @@ func (c *Client) ModuleLocation(ctx context.Context, module *regsrc.Module, vers
 		}
 
 		location = v.Location
+
+		// if the location is empty, we will fallback to the header
+		if location == "" {
+			location = resp.Header.Get(xTerraformGet)
+		}
 
 	case http.StatusNoContent:
 		// FALLBACK: set the found location from the header

